stingray/pulse/pulsar.py

```diff
@@ -313,31 +414,201 @@ def z_n(phase, n=2, norm=1):
     n : int, default 2
         Number of harmonics, including the fundamental
 
+    Returns
+    -------
+    ks : list of ints
+        Harmonic numbers, from 1 to nmax (included)
+    z2_n : float
+        The Z^2_n statistic for all ks
+    '''
+    all_zs = np.zeros(nmax)
+    ks = np.arange(1, nmax + 1)
+    nphot = phase.size
+
+    total_sum = 0
+    phase = phase * 2 * np.pi
+
+    for k in ks:
+        local_z = (
+            np.sum(np.cos(k * phase)) ** 2
+            + np.sum(np.sin(k * phase)) ** 2
+        )
+        total_sum += local_z
+        all_zs[k - 1] = total_sum
+
+    return ks, 2 / nphot * all_zs
+
+
+def z_n_poisson(profile, n):
+    '''Z^2_n statistic for Poissonian pulse profiles
+
+    See Bachetti+2021, arXiv:2012.11397
+
+    Parameters
+    ----------
+    profile : array of floats
+        The folded pulse profile (containing the number of
+        photons falling in each pulse bin)
+    n : int
+        Number of harmonics, including the fundamental
+
+    Returns
+    -------
+    z2_n : float
+        The value of the statistic
+    '''
+    _, all_zs = z_n_poisson_all(profile, nmax=n)
+    return all_zs[-1]
+
+
+def z_n_gauss(profile, err, n):
+    '''Z^2_n statistic for Gaussian pulse profiles
+
+    See Bachetti+2021, arXiv:2012.11397
+
+    Parameters
+    ----------
+    profile : array of floats
+        The folded pulse profile
+    err : float
+        The (assumed constant) uncertainty on the flux in each bin.
+    n : int
+        Number of harmonics, including the fundamental
+
+    Returns
+    -------
+    z2_n : float
+        The value of the statistic
+    '''
+    _, all_zs = z_n_gauss_all(profile, err, nmax=n)
+    return all_zs[-1]
+
+
+def z_n_events(phase, n):
+    '''Z^2_n statistics, a` la Buccheri+03, A&A, 128, 245, eq. 2.
+
+    Parameters
+    ----------
+    phase : array of floats
+        The phases of the events
+    n : int, default 2
+        Number of harmonics, including the fundamental
+
+    Returns
+    -------
+    z2_n : float
+        The Z^2_n statistic
+    '''
+    ks, all_zs = z_n_events_all(phase, nmax=n)
+    return all_zs[-1]
+
+
+def z_n(data, n, kind="events", err=None, norm=None):
```

Comment:
suggestion: rename "kind" to "datatype" to be more descriptive.

Reply:
Done